lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -809,6 +813,10 @@ private function computeAssociationChanges($assoc, $value)
         $targetClass    = $this->em->getClassMetadata($assoc['targetEntity']);
 
         foreach ($unwrappedValue as $key => $entry) {
+            if (! ($entry instanceof $assoc['targetEntity']))) {
+                throw ORMInvalidArgumentException::invalidAssociation($entry);
```

Comment:
You should throw with `$targetClass`, `$assoc`, the original entity metadata and the `$entry`: that allows you to build a more specific error message.